### ISSUE TITLE
Terraform 1.6 Crash

### ISSUE THREAD

**Issue by DenisBalan** [hashicorp/terraform#34024]
### Terraform Version

```shell
/usr/local/bin/terraform version
Terraform v1.6.0
on linux_amd64
```

### Terraform Configuration Files

```terraform
...terraform config...
```

### Debug Output

2023-10-09T10:54:30.5086654Z [command]/usr/local/bin/terraform version
2023-10-09T10:54:30.5812003Z Terraform v1.6.0
2023-10-09T10:54:30.5812584Z on linux_amd64
2023-10-09T10:54:30.5813519Z + provider registry.terraform.io/azure/azapi v1.9.0
2023-10-09T10:54:30.5813913Z + provider registry.terraform.io/hashicorp/azuread v2.43.0
2023-10-09T10:54:30.5814338Z + provider registry.terraform.io/hashicorp/azurerm v3.75.0
2023-10-09T10:54:30.5814994Z + provider registry.terraform.io/hashicorp/random v3.5.1
2023-10-09T10:54:30.5815373Z + provider registry.terraform.io/hashicorp/tls v4.0.4
2023-10-09T10:54:30.5966494Z [command]/usr/local/bin/terraform plan -var-file=../env/.../terraform.tfvars -input=false -compact-warnings -detailed-exitcode -out=tfplan.zip
2023-10-09T10:54:38.7080247Z 
2023-10-09T10:54:38.7081147Z !!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
2023-10-09T10:54:38.7081593Z 
2023-10-09T10:54:38.7082002Z Terraform crashed! This is always indicative of a bug within Terraform.
2023-10-09T10:54:38.7082930Z Please report the crash with Terraform[1] so that we can fix this.
2023-10-09T10:54:38.7083144Z 
2023-10-09T10:54:38.7083345Z When reporting bugs, please include your terraform version, the stack trace
2023-10-09T10:54:38.7083594Z shown below, and any additional information which may help replicate the issue.
2023-10-09T10:54:38.7083703Z 
2023-10-09T10:54:38.7083868Z [1]: https://github.com/hashicorp/terraform/issues
2023-10-09T10:54:38.7083953Z 
2023-10-09T10:54:38.7084139Z !!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
2023-10-09T10:54:38.7084245Z 
2023-10-09T10:54:38.7084398Z value is marked, so must be unmarked first
2023-10-09T10:54:38.7389649Z goroutine 2154 [running]:
2023-10-09T10:54:38.7390408Z runtime/debug.Stack()
2023-10-09T10:54:38.7390791Z 	/opt/hostedtoolcache/go/1.21.1/x64/src/runtime/debug/stack.go:24 +0x5e
2023-10-09T10:54:38.7391200Z runtime/debug.PrintStack()
2023-10-09T10:54:38.7391529Z 	/opt/hostedtoolcache/go/1.21.1/x64/src/runtime/debug/stack.go:16 +0x13
2023-10-09T10:54:38.7392244Z github.com/hashicorp/terraform/internal/logging.PanicHandler()
2023-10-09T10:54:38.7392596Z 	/home/runner/work/terraform/terraform/internal/logging/panic.go:58 +0x13b
2023-10-09T10:54:38.7392934Z panic({0x2b92ec0?, 0x38c4c30?})
2023-10-09T10:54:38.7393271Z 	/opt/hostedtoolcache/go/1.21.1/x64/src/runtime/panic.go:920 +0x270
2023-10-09T10:54:38.7393973Z github.com/zclconf/go-cty/cty.Value.assertUnmarked(...)
2023-10-09T10:54:38.7394455Z 	/home/runner/go/pkg/mod/github.com/zclconf/go-cty@v1.14.0/cty/marks.go:141
2023-10-09T10:54:38.7395007Z github.com/zclconf/go-cty/cty.Value.AsString({{{0x3902720?, 0xc000146bd1?}}, {0x2e1de40?, 0xc002d08270?}})
2023-10-09T10:54:38.7395536Z 	/home/runner/go/pkg/mod/github.com/zclconf/go-cty@v1.14.0/cty/value_ops.go:1385 +0x47
2023-10-09T10:54:38.7396081Z github.com/zclconf/go-cty/cty.Value.Range({{{0x3902720?, 0xc000146bd1?}}, {0x2e1de40?, 0xc002d08270?}})
2023-10-09T10:54:38.7396854Z 	/home/runner/go/pkg/mod/github.com/zclconf/go-cty@v1.14.0/cty/value_range.go:53 +0x285
2023-10-09T10:54:38.7397286Z github.com/hashicorp/hcl/v2/hclsyntax.(*ConditionalExpr).Value(0xc00051ac40, 0xc002d08240)
2023-10-09T10:54:38.7397705Z 	/home/runner/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.18.0/hclsyntax/expression.go:746 +0xd6a
2023-10-09T10:54:38.7398147Z github.com/hashicorp/hcl/v2/hcldec.(*AttrSpec).decode(0xc002d11590, 0xc001beccf8?, {0xc00162aaa0?, 0xe?, 0xc000cf3650?}, 0xc002d08240)
2023-10-09T10:54:38.7398584Z 	/home/runner/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.18.0/hcldec/spec.go:214 +0x1f5
2023-10-09T10:54:38.7399006Z github.com/hashicorp/hcl/v2/hcldec.ObjectSpec.decode(0xc002d10f90, 0xc002d10f90?, {0x0, 0x0, 0x0}, 0xc002ddcae0?)
2023-10-09T10:54:38.7399419Z 	/home/runner/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.18.0/hcldec/spec.go:88 +0x1f7
2023-10-09T10:54:38.7399839Z github.com/hashicorp/hcl/v2/hcldec.decode({0x3903ae0, 0xc000927400}, {0x0, 0x0, 0x0}, 0x0?, {0x3902fb8, 0xc002d10f90}, 0x0)
2023-10-09T10:54:38.7400270Z 	/home/runner/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.18.0/hcldec/decode.go:24 +0x10f
2023-10-09T10:54:38.7400625Z github.com/hashicorp/hcl/v2/hcldec.Decode(...)
2023-10-09T10:54:38.7400963Z 	/home/runner/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.18.0/hcldec/public.go:18
2023-10-09T10:54:38.7401382Z github.com/hashicorp/terraform/internal/lang.(*Scope).EvalBlock(0xc001e0eab0, {0x3903530, 0xc00267da10}, 0xc002d75178?)
2023-10-09T10:54:38.7401776Z 	/home/runner/work/terraform/terraform/internal/lang/eval.go:71 +0x1ea
2023-10-09T10:54:38.7402238Z github.com/hashicorp/terraform/internal/terraform.(*BuiltinEvalContext).EvaluateBlock(0xc000870a20?, {0x3902d18, 0xc0010456b0}, 0xc001344360?, {0x0?, 0x0?}, {{{{0x0, 0x0}}, {0x0, 0x0}}, ...})
2023-10-09T10:54:38.7402724Z 	/home/runner/work/terraform/terraform/internal/terraform/eval_context_builtin.go:282 +0x145
2023-10-09T10:54:38.7403183Z github.com/hashicorp/terraform/internal/terraform.(*NodeValidatableResource).validateResource(0xc00087ff08, {0x391fb60, 0xc002c6fce0})
2023-10-09T10:54:38.7403645Z 	/home/runner/work/terraform/terraform/internal/terraform/node_resource_validate.go:372 +0x666
2023-10-09T10:54:38.7404096Z github.com/hashicorp/terraform/internal/terraform.(*NodeValidatableResource).Execute(0xc00087ff08, {0x391fb60, 0xc002c6fce0}, 0x1?)
2023-10-09T10:54:38.7404541Z 	/home/runner/work/terraform/terraform/internal/terraform/node_resource_validate.go:52 +0x55
2023-10-09T10:54:38.7405005Z github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc003a8db00, {0x391fb60, 0xc002c6fce0}, {0x7ff8e3e56380, 0xc00087ff08})
2023-10-09T10:54:38.7405473Z 	/home/runner/work/terraform/terraform/internal/terraform/graph_walk_context.go:143 +0xbe
2023-10-09T10:54:38.7405887Z github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x30d81c0, 0xc00087ff08})
2023-10-09T10:54:38.7406283Z 	/home/runner/work/terraform/terraform/internal/terraform/graph.go:78 +0x375
2023-10-09T10:54:38.7406709Z github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc0014353e0, {0x30d81c0, 0xc00087ff08}, 0xc0013409c0)
2023-10-09T10:54:38.7407206Z 	/home/runner/work/terraform/terraform/internal/dag/walk.go:384 +0x2e5
2023-10-09T10:54:38.7407590Z created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 10
2023-10-09T10:54:38.7407973Z 	/home/runner/work/terraform/terraform/internal/dag/walk.go:307 +0xde8
2023-10-09T10:54:38.7490913Z ##[error]Terraform command 'plan' failed with exit code '11'.
2023-10-09T10:54:38.7596028Z ##[error]
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

value is marked, so must be unmarked first
goroutine 2154 [running]:
runtime/debug.Stack()
	/opt/hostedtoolcache/go/1.21.1/x64/src/runtime/debug/stack.go:24 +0x5e
runtime/debug.PrintStack()
	/opt/hostedtoolcache/go/1.21.1/x64/src/runtime/debug/stack.go:16 +0x13
github.com/hashicorp/terraform/internal/logging.PanicHandler()
	/home/runner/work/terraform/terraform/internal/logging/panic.go:58 +0x13b
panic({0x2b92ec0?, 0x38c4c30?})
	/opt/hostedtoolcache/go/1.21.1/x64/src/runtime/panic.go:920 +0x270
github.com/zclconf/go-cty/cty.Value.assertUnmarked(...)
	/home/runner/go/pkg/mod/github.com/zclconf/go-cty@v1.14.0/cty/marks.go:141
github.com/zclconf/go-cty/cty.Value.AsString({{{0x3902720?, 0xc000146bd1?}}, {0x2e1de40?, 0xc002d08270?}})
	/home/runner/go/pkg/mod/github.com/zclconf/go-cty@v1.14.0/cty/value_ops.go:1385 +0x47
github.com/zclconf/go-cty/cty.Value.Range({{{0x3902720?, 0xc000146bd1?}}, {0x2e1de40?, 0xc002d08270?}})
	/home/runner/go/pkg/mod/github.com/zclconf/go-cty@v1.14.0/cty/value_range.go:53 +0x285
github.com/hashicorp/hcl/v2/hclsyntax.(*ConditionalExpr).Value(0xc00051ac40, 0xc002d08240)
	/home/runner/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.18.0/hclsyntax/expression.go:746 +0xd6a
github.com/hashicorp/hcl/v2/hcldec.(*AttrSpec).decode(0xc002d11590, 0xc001beccf8?, {0xc00162aaa0?, 0xe?, 0xc000cf3650?}, 0xc002d08240)
	/home/runner/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.18.0/hcldec/spec.go:214 +0x1f5
github.com/hashicorp/hcl/v2/hcldec.ObjectSpec.decode(0xc002d10f90, 0xc002d10f90?, {0x0, 0x0, 0x0}, 0xc002ddcae0?)
	/home/runner/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.18.0/hcldec/spec.go:88 +0x1f7
github.com/hashicorp/hcl/v2/hcldec.decode({0x3903ae0, 0xc000927400}, {0x0, 0x0, 0x0}, 0x0?, {0x3902fb8, 0xc002d10f90}, 0x0)
	/home/runner/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.18.0/hcldec/decode.go:24 +0x10f
github.com/hashicorp/hcl/v2/hcldec.Decode(...)
	/home/runner/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.18.0/hcldec/public.go:18
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalBlock(0xc001e0eab0, {0x3903530, 0xc00267da10}, 0xc002d75178?)
	/home/runner/work/terraform/terraform/internal/lang/eval.go:71 +0x1ea
github.com/hashicorp/terraform/internal/terraform.(*BuiltinEvalContext).EvaluateBlock(0xc000870a20?, {0x3902d18, 0xc0010456b0}, 0xc001344360?, {0x0?, 0x0?}, {{{{0x0, 0x0}}, {0x0, 0x0}}, ...})
	/home/runner/work/terraform/terraform/internal/terraform/eval_context_builtin.go:282 +0x145
github.com/hashicorp/terraform/internal/terraform.(*NodeValidatableResource).validateResource(0xc00087ff08, {0x391fb60, 0xc002c6fce0})
	/home/runner/work/terraform/terraform/internal/terraform/node_resource_validate.go:372 +0x666
github.com/hashicorp/terraform/internal/terraform.(*NodeValidatableResource).Execute(0xc00087ff08, {0x391fb60, 0xc002c6fce0}, 0x1?)
	/home/runner/work/terraform/terraform/internal/terraform/node_resource_validate.go:52 +0x55
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc003a8db00, {0x391fb60, 0xc002c6fce0}, {0x7ff8e3e56380, 0xc00087ff08})
	/home/runner/work/terraform/terraform/internal/terraform/graph_walk_context.go:143 +0xbe
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x30d81c0, 0xc00087ff08})
	/home/runner/work/terraform/terraform/internal/terraform/graph.go:78 +0x375
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc0014353e0, {0x30d81c0, 0xc00087ff08}, 0xc0013409c0)
	/home/runner/work/terraform/terraform/internal/dag/walk.go:384 +0x2e5
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 10
	/home/runner/work/terraform/terraform/internal/dag/walk.go:307 +0xde8

2023-10-09T10:54:38.8445945Z ##[section]Finishing: terraform plan

### Expected Behavior

Should work with no issues, as it is on `1.5.7`

### Actual Behavior

Crash, as per debug error message

### Steps to Reproduce

terraform init
terraform plan

### Additional Context

It is working fine on `1.5.7`, after last update to `1.6` it is crashing completly.
No other changes were made.

On `Sep 28, 2023`
```
/usr/local/bin/terraform version
Terraform v1.5.7
on linux_amd64
```
Everything is fine.

No code changes at all!
Microsoft Azure hosted agent `ubuntu-latest` updated Terraform version to `1.6`, now same code is crashing with error mentioned.

Any ideas?

### References

#34022

**Comment by DenisBalan** [comment]
Another issue with same erorr message #34023

**Comment by liamcervante** [comment]
Thanks for reporting this!

I believe this is a duplicate of https://github.com/hashicorp/terraform/issues/33977, and a fix has already been merged for release in `v1.6.1`
